### ISSUE TITLE
Corrupted payload with uSockWrite

### ISSUE THREAD

**Issue by benhaub** [u-blox/ubxlib#156]
## Environment
Module type: SARA-U201
Device type: Cell
Network type: Cell
Enable power pin: N/C (tied to 3v3)
Power-on pin: N/C tied high
VInt pin: N/C
DTR power saving pin: N/C
Transport type: UART
Baud rate: 115200
CTS/RTS: N/C, pulled to ground with 2kΩ resistors

Platform: esp-idf
Firmware: esp-idf v5.1.1
ubxlib: bd8a69efdecc6655fd2f6fcaebc9d63dac1f7571
Example: main_tls.c with no TLS, no credential checking.
## Issue

When opening a TCP cellular socket, the payload data is corrupted by the AT+USOWR command (i.e. the beginning of my payload is corrupted and an AT+USOWR is appended and sent to my http server which corrupts my request).

I have a mod on my board which pulls CTS and RTS down to ground with 2kΩ resistors since they are not connected and serial communication over UART was not possible with those pins floating.

See similar issue here:
https://portal.u-blox.com/s/question/0D52p00009JhO5VCAV/sara-r412m-sending-at-command-on-socket

## Workaround
To fix this issue I had to disable UART power saving for modules that don't use CTS/RTS and use Tx data to wakeup by commenting out the following:
```
//else clause in u_cell_pwr.c:987
else {
            success = moduleConfigureOne(atHandle, "AT&K0",
                                         U_CELL_PWR_CONFIGURATION_COMMAND_TRIES);
            // RTS/CTS handshaking is not used by the UART HW, we
            // can use the wake-up on TX line feature without any
            // complications
            if (uAtClientWakeUpHandlerIsSet(atHandle) &&
                U_CELL_PRIVATE_HAS(pInstance->pModule,
                                   U_CELL_PRIVATE_FEATURE_UART_POWER_SAVING)) {
                uartPowerSavingMode = U_CELL_PWR_PSV_MODE_DATA;
            }
```
 (i.e. do not send AT+UPSV=1,1300, or likely any other derivative of that command).

**Comment by RobMeades** [comment]
Hi, and thanks very much for the detailed post.  Let me replay what you've said to see if I've understood it correctly.

1.  UART power saving is on (it will be by default) and is set up to wake the module on TX line activity with a 6-second (1300 GSM frames) timeout, also the default.
2. You have disabled CTS and RTS and tied both to ground.
3. Under these circumstances, data corruption occurs when you use the sockets interface.
4. An aspect of this corruption is that a bit of the next AT command ends up appended to the data you are sending.

Is that a correct interpretation?

First comment is that, it is probably not well documented but, you might be able to just define [U_CFG_CELL_DISABLE_UART_POWER_SAVING](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/src/u_cell.c#L414) for your build, to avoid needing to comment out any code.

It seems likely that what is happening is that, somehow, the module is not properly awake after it has sent the `@` character that prompts the `ubxlib` code for the data.  Since the AT command that prompted the `@` character said "I'm gonna send you this many characters", if it somehow misses characters at the start it will just pick them up from the thing that follows, which happens to be the next AT command, which will of course mess that one up also, all not good.

First question, I guess, is whether power saving is important to you.  If not you could hopefully just get on with life by either employing your workaround or defining `U_CFG_CELL_DISABLE_UART_POWER_SAVING` for your build.

If power saving is important to you, do you happen to have a Saleae probe or similar and could capture the problem occurring on the UART interface with that?  If we can see the exact timing we can likely figure out which bit is not awake when and mitigate this somehow in the sockets code.

**Comment by benhaub** [comment]
>Is that a correct interpretation?

Everything is correct except for the 4th point.

The behaviour I see is that it's actually the _previous_ AT command being sent along with my payload. I see AT+USOWR=sock,bytesToWrite, **not** +USOWR: sock,bytesWritten. The difference there being is that the previous command is on the Tx line along with my payload, while the next command is on the Rx line.

>is power saving is important to you

Yeah it is important. My hardware team has left the CTS and RTS pins not connected to save on pin count so we can scope it to see what's going on.

**Comment by RobMeades** [comment]
Thanks for the swift response.  From a terminology point of view, I think taken from the ITU V.250 standard, the `AT+BLAH` thing would usually be referred to as the "command" and the "+BLAH:" thing that follows the "information response".

To clarify then, if you were sending the payload "The quick brown fox jumps over the lazy dog":

```
AT+USOWR=0,44
@The quick brown fox jumps over the lazy dog.
+USOWR: 0,44
OK
```

...what ends up at the far end is:

<corruption of some form><some portion of the intended payload><some portion of `AT+USOWR=0,44`>

Is there any chance that the bit at the end, the "some portion of `AT+USOWR=sock,bytesToWrite`, is from the _next_ write to the socket?  i.e. if sending in a sequence, might it be:

```
AT+USOWR=0,44
@The quick brown fox jumps over the lazy dog.
+USOWR: 0,44
OK
AT+USOWR=...   // This bit appended to the end of the intended message?
```

Maybe paste in here a segment of AT log, to make it clear what's going on?

**Comment by benhaub** [comment]
>defining U_CFG_CELL_DISABLE_UART_POWER_SAVING for your build.

Tried this but the power saving configuration is set based upon enabled pins within `moduleConfigure`, which does not call `uCellPwrEnableUartSleep`, at least in the cell API.

>[corruption of some form][some portion of the intended payload][some portion of AT+USOWR=0,44]

That is completely correct. Exactly.

>Is there any chance that the bit at the end, the "some portion of AT+USOWR=sock,bytesToWrite, is from the next write to the socket?

No. I send a single packet with a payload of 137 bytes..

Here is some screenshots. I can see some slightly different behaviour when I change my message from an http request to "The quick brown fox jumps over the lazy dog" in that now the beginning of the payload is _corrupted and replaced_ with the AT command where as before the payload was _corrupted and appended_ with an AT command.

ubxlib AT log:
![Screenshot 2023-11-02 at 3 24 19 PM](https://github.com/u-blox/ubxlib/assets/39037440/e1d7e5b6-0ae5-4bd3-b5b3-eb8d7793583a)

Wireshark:
![Screenshot 2023-11-02 152942](https://github.com/u-blox/ubxlib/assets/39037440/0ff259e2-5a40-45f5-9e4c-79d8c1feaff3)

My server:
![Screenshot 2023-11-02 152942](https://github.com/u-blox/ubxlib/assets/39037440/57d7a995-e112-4e92-852a-1a7e369b2d52)

**Comment by RobMeades** [comment]
Very interesting: so the payload is 44 bytes long (including the null terminator):

"The quick brown fox jumps over the lazy dog[00]"

Assuming that what you have highlighted in the Wireshark log is the payload, then there are:

- 7 bytes of corruption: `Y[8a][d5][a5][8d][ad][81]`,
- the first 10 bytes of the intended payload ("`The quick `") are missing,
- the remaining 34 bytes of the intended payload ("`brown ... dog[00]`") are received correctly,
- 3 bytes, `[0d]AT`, are added at the end, where I think the linefeed character, `0x0d`, results in your print making it look like they are at the front.
- this makes a total of 7 + 34 + 3 = 44 bytes of payload; nice that something adds up.

Of course the characters `[0d]AT` are visible in the `ubxlib` log.  The only reason they would be sent by `ubxlib`, and be included in the payload, is if the module was sitting there waiting for more characters: at the same time `ubxlib` is waiting for the module to return `OK` and it hasn't so `ubxlib` is prodding it to make it do something.  This is consistent with the module having missed characters from the start of the sequence: it wants 44 of them and hasn't got 44 of them.  If `[0d]AT` were enough to make the module transmit the payload we can guess that it had only received 41 of the 44 bytes that it was sent.

I've no idea why that might be but, purely as an experiment, would you be able to try changing the value here:

https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/src/u_cell_sock.c#L1591

....to be, say, 100, to see if that has any effect on the world?  To be clear, this is not a solution, just an investigation.

**Comment by benhaub** [comment]
Seems like it made it better, but not quite fixing it.
100ms:
![100delay](https://github.com/u-blox/ubxlib/assets/39037440/10da071a-c7c1-49d8-af72-bf72166376d1)

200ms
![200delay](https://github.com/u-blox/ubxlib/assets/39037440/e7e5a9f9-19f3-4e71-8ec1-572f44f93e46)

400ms
![400delay](https://github.com/u-blox/ubxlib/assets/39037440/9e207752-ff50-4379-8ecb-d2224251d444)

1200ms
![1200 delay](https://github.com/u-blox/ubxlib/assets/39037440/48c56125-6d39-4b2e-9214-704544e94f2c)

**Comment by RobMeades** [comment]
On the `U_CFG_CELL_DISABLE_UART_POWER_SAVING`, checking if I've misunderstood:

- the code [here](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/src/u_cell.c#L414), which would be omitted if `U_CFG_CELL_DISABLE_UART_POWER_SAVING` were defined, is the code that sets `uCellPrivateWakeUpCallback()` as a wake-up handler inside the AT Client; so if `U_CFG_CELL_DISABLE_UART_POWER_SAVING` is defined for the build, no wake-up handler should be set inside the AT Client,
- this handler, inside the AT Client, is the thing that sends that `AT` we talked about above if there has been no AT command for, in this case, for 6 seconds: this is required, when UART power saving, is active to wake the module up.
- the code over in [moduleConfigure()](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/src/u_cell_pwr.c#L966) checks whether such a wake-up handler has been set in the AT Client by calling `uAtClientWakeUpHandlerIsSet()`; it does this in both halves of the `if()` check concerning whether the flow control lines are on or not, and in most of the other `if()` checks that follow; if no AT Client wake-up handler has been set this should always return `false`.

Basically, the intention is that if that handler is not set no power saving of any form should occur.  Can you expand a little on why this does not have the desired effect for you?

**Comment by RobMeades** [comment]
> Seems like it made it better, but not quite fixing it.

Thanks for doing that and, yes, it is utterly mad.  Even assuming that UART power saving were enabled, the wake-up from 32 kHz sleep inside the module takes a few milliseconds, not half a second.

Could you possibly paste into here a log of all the rubbish that `ubxlib` spits out, from when it starts up, in case anything is going wrong with any of the configuration steps?

**Comment by benhaub** [comment]
```
U_CELL: initialising with enable power pin not connected, PWR_ON pin not connected and VInt pin not connected.
[00]AT
AT
OK
U_CELL_PWR: powering on, module is already on.
ATE0
ATE0
OK
AT+CMEE=2
OK
AT+UDCONF=1,0
+UMWI: 0,1
+UMWI: 0,2
+UMWI: 0,3
+UMWI: 0,4
OK
ATI9
23.60,A01.01
OK
AT&C1
OK
AT&D1
OK
AT&K0
OK
AT+UPSV=1,1300
OK
AT+UGPRF?
AT
AT
OK
AT+CFUN=0
OK
Opened device with return code 0.
Bringing up the network...
U_CELL_NET: preparing to register/connect...
AT+CREG=2
OK
AT+CGREG=2
OK
AT+CIMI
234500023270006
OK
U_CELL_NET: user-specified APN is "jtm2m".
U_CELL_NET: setting automatic network selection mode...
AT+COPS?
+COPS: 0
OK
AT+CFUN=1
OK
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
AT+CGREG?
+CGREG: 2,0
OK
0: NReg
AT+CREG?
+CREG: 2,0
OK
0: NReg
+CREG: 5,"EA6D","544C28",2
+CGREG: 5,"EA6D","544C28",2,"01"
2: RegR
2: RegR
AT+COPS=3,0
OK
AT+COPS?
+COPS: 0,0,"Rogers Wireless",2
OK
AT+CGATT?
+CGATT: 1
OK
AT+UPSD=0,1,"jtm2m"
OK
AT+UPSD=0,2,""
OK
AT+UPSD=0,3,""
OK
AT+UPSD=0,7,"0.0.0.0"
OK
AT+UPSD=0,6,3
OK
AT+UPSDA=0,3
OK
U_CELL_NET: connected after 12 second(s).
Looking up server address...
Address is: IPV4 38.23.179.39
Creating socket...
AT+USOCR=6
+USOCR: 0
OK
U_SOCK: socket created, descriptor 0, network handle 0x3fcc0b18, socket handle 0.
U_SOCK: connecting socket to "38.23.179.39:8080"...
AT+USOCO=0,"38.23.179.39",8080
OK
U_SOCK: socket with descriptor 0, network handle 0x3fcc0b18, socket handle 0, is  connected to address "38.23.179.39:8080".
Sending data...
AT+USOWR=0,44
@The quick brown fox jumps over the lazy dog[00]AT
AT
+USOWR: 0,44
OK
AT+USOER
+USOER: 0
OK
Sent 44 byte(s) to echo server.
Done.
```

**Comment by RobMeades** [comment]
That looks quite sensible, UART power saving is enabled if nothing happens for 6 seconds.  What we appear to be _seeing_, however, is only the periodic wake-up on paging and NOT the wake-up on a character arriving at the module.

FYI, integration manual:

https://content.u-blox.com/sites/default/files/SARA-G3-U2_SysIntegrManual_UBX-13000995.pdf

...page 58, "Wake-up via data reception", is the detailed reference.

It says:

> On SARA-U2 series, the TXD input line is configured to wake up the system via data reception if `AT+UPSV=1` is set with hardware flow control disabled

We have that, `AT&K0` being the bit that disables the flow control.

I will continue thinking...

**Comment by RobMeades** [comment]
If you get a moment it might be worth adding a call to [uAtClientTimestampSet(0)](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/common/at_client/api/u_at_client.h#L632), just after you have brought the cellular device up, and paste here another log of the bit around the socket send.

This will put timestamps into the AT prints, which might be useful.

**Comment by benhaub** [comment]
```
U_CELL_NET: connected after 11 second(s).
Looking up server address...
Address is: IPV4 38.23.179.39
Creating socket...
0000/01/01 00:00:00.049: AT+USOCR=6
0000/01/01 00:00:00.089:
0000/01/01 00:00:00.089: +USOCR: 0
0000/01/01 00:00:00.090:
0000/01/01 00:00:00.090: OK
U_SOCK: socket created, descriptor 0, network handle 0x3fcc0b18, socket handle 0.
U_SOCK: connecting socket to "38.23.179.39:8080"...
0000/01/01 00:00:00.149: AT+USOCO=0,"38.23.179.39",8080
0000/01/01 00:00:00.529:
0000/01/01 00:00:00.529: OK
U_SOCK: socket with descriptor 0, network handle 0x3fcc0b18, socket handle 0, is  connected to address "38.23.179.39:8080".
Sending data...
0000/01/01 00:00:00.589: AT+USOWR=0,44
0000/01/01 00:00:00.639:
0000/01/01 00:00:00.639: @0000/01/01 00:00:00.689: The quick brown fox jumps over the lazy dog[00]AT
0000/01/01 00:00:08.759: AT
0000/01/01 00:00:08.919:
0000/01/01 00:00:08.919: +USOWR: 0,44
0000/01/01 00:00:08.920:
0000/01/01 00:00:08.920: OK
0000/01/01 00:00:08.920: AT+USOER
0000/01/01 00:00:08.959:
0000/01/01 00:00:08.959: +USOER: 0
0000/01/01 00:00:08.959:
0000/01/01 00:00:08.960: OK
Sent 44 byte(s) to echo server.
Done.
```

**Comment by RobMeades** [comment]
Just weird.  The module must have been awake at 0.589 seconds, because it reacted to the `AT+USOWR=0,44` and sent the `@` prompt that arrived 50 milliseconds later and that we reacted to by sending the payload 50 milliseconds after that.

Losing the first 7 characters of the payload to garbage makes no sense, I can't see how it can be down to power saving.  Yet your experiments show that setting `UPSV=0` stops the problem occurring.

Hmph.

**Comment by benhaub** [comment]
Yeah just, for your sanity, here is a working example where this time the power saving command was `AT+UPSV=0` since `U_CFG_CELL_DISABLE_UART_POWER_SAVING` was defined.

Wireshark:
![Screenshot 2023-11-02 152942](https://github.com/u-blox/ubxlib/assets/39037440/77685dd2-8329-4dff-b34e-766176728572)

My server:
![Screenshot 2023-11-02 152942](https://github.com/u-blox/ubxlib/assets/39037440/59fc1383-9fba-4acd-a83c-a88ac31149d3)

Ubxlib AT log
![Screenshot 2023-11-02 at 5 45 26 PM](https://github.com/u-blox/ubxlib/assets/39037440/400328a9-e2dc-442b-9b26-81908d2b23f8)

**Comment by RobMeades** [comment]
I'm sane!  Well, at least, thank you for the straw to clutch at.  Having a working thing is a good start.

FYI, the main support guys for SARA-U201 are in Italy which is currently experiencing one of those European national holidays that are on a Thursday and so no-one will be in now until next week.  If nothing has occurred to me by then I know who to contact though.

**Comment by RobMeades** [comment]
OK: do you want the bad news or the bad news [but, don't worry, there is a way forward at the end :-)]?

First thing, in case you weren't aware, is that, due to lack of availability of a critical component, SARA-U201 was "end of lifed" a little while ago:

https://www.u-blox.com/sites/default/files/SARA-U201_EOL_UBX-22002617.pdf 

In fact we were planning on beginning deprecation of it in `ubxlib` from our next release (1.4, due start 2024), eventually dropping support for it likely mid next year, though if people are still using it we will not do that.

The reason I mention this is that a thorough search of our internal JIRA database reveals that there is a bug in SARA-U201 which means that, when UART power saving is enabled, it immediately goes into 32 kHz sleep, if it can, after it has sent the `@` prompt.  And because the product is end of life this was not fixed. Crazy, I know, but seems to be the case.

There are two workarounds for this:

- call [uCellSockHexModeOn()](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/api/u_cell_sock.h#L296) to use hex mode instead of binary mode; in this case there is no `@` prompt, so no `@` prompt problem, but it will halve the amount of data that can go into each call to [uCellSockWrite()](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/src/u_cell_sock.c#L1530); doesn't halve the throughput or anything, just the number of calls you might need to make if the amount of data is > 512ish bytes.
- hack something like this into both [uCellSockWrite()](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/src/u_cell_sock.c#L1530) and probably also [uCellSockSendTo()](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/src/u_cell_sock.c#L1230C9-L1230C24) (for the UDP case):

```
uAtClientLock(atHandle);  // Existing AT Client lock function

// Disable UART power saving while doing the @ process
uAtClientCommandStart(atHandle, "AT+UPSV=0");
uAtClientCommandStopReadResponse(atHandle);

uAtClientCommandStart(atHandle, "AT+USxxx=");  // Existing USOWR/USOST command

...  // Send etc. is performed

// At the end, re-enable UART power saving
uAtClientCommandStart(atHandle, "AT+UPSV=1,1300");
uAtClientCommandStopReadResponse(atHandle);

uAtClientUnloock(atHandle);  // Existing AT Client unlock function
```

Taking the second approach will slow things down somewhat, an additional pair of AT exchanges per call to [uCellSockWrite()](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/src/u_cell_sock.c#L1530)/[uCellSockSendTo()](https://github.com/u-blox/ubxlib/blob/cd01364006c3bcf1476b7711ae43b5fbf7b4a79b/cell/src/u_cell_sock.c#L1230C9-L1230C24), doing the first doesn't have any significant drawbacks if you're only using TCP; for UDP it has an impact since the maximum UDP packet size is reduced, but since TCP is a stream you will be unlikely to notice anything.

An additional question: what is your "usage scenario"?  Is it one where the device will sit in idle, transmitting, say, every few minutes/10s of minutes, maybe waiting for stuff to arrive on the downlink of an established TCP socket, or is it one where the device will power up, register, do something, and then switch off again to save power?  Fundamentally, what is the periodicity of data transmission and how much dependency is there on downlink stuff?  I ask because, if your use-case falls more into the second camp then UART power saving isn't going to save you very much anyway, you will burn the vast majority of your power waking up and registering; UART power saving is only helpful in the first camp.

**Comment by benhaub** [comment]
It was some planned testing I have to do.

**Comment by benhaub** [comment]
>First thing, in case you weren't aware, is that, due to lack of availability of a critical component, SARA-U201 was "end of lifed" a little while ago:

I did not know that but I think we'll be using them for a little while longer considering what we have in inventory.
>There are two workarounds for this

Thanks. Either one of those should work. We have quite a bit of unused cycles on our s3.
>what is your "usage scenario"?

Right now we send data at 1Hz, but I had planned to see what the power savings would be if that were decreased to 0.5Hz or 0.2Hz.

**Comment by RobMeades** [comment]
> Right now we send data at 1Hz, but I had planned to see what the power savings would be if that were decreased to 0.5Hz or 0.2Hz.

You will be remaining registered for these kinds of rates, not powering off, so agreed that `AT+UPSV` is the thing.

> Either one of those should work. We have quite a bit of unused cycles on our s3.

I'd go for the "hex mode" approach, there should be no down-sides, you can still use `AT+UPSV` and you don't have to make any hacky code modifications.

> we'll be using them for a little while longer considering what we have in inventory.

In that case we won't deprecate SARA-U201 just yet.  Note, though, that we no longer actively test `ubxlib` with a SARA-U201: this is because the UK networks that we are using, for 2G, now implement the awful/stupid GSMA radio limitations which mean that anything that powers on and off again frequently, which we inevitably do during testing, gets banned from the network for an hour, making meaningful test operation impossible.  All _should_ still be fine (we test everything on Cat-M1/LTE RAT), please let us know if you come up against any problems and we will address them.

**Comment by benhaub** [comment]
Thanks Rob, will go with hex mode.

**Comment by RobMeades** [comment]
Good stuff, gonna close this one, please re-open or open a new one as necessary.
